File: src/main/scala/li/cil/oc/common/tileentity/traits/OpenSides.scala
package li.cil.oc.common.tileentity.traits

import li.cil.oc.Settings
import li.cil.oc.util.RotationHelper
import net.minecraft.nbt.CompoundNBT
import net.minecraft.util.Direction
import net.minecraftforge.api.distmarker.Dist
import net.minecraftforge.api.distmarker.OnlyIn

/**
  * @author Vexatos
  */
trait OpenSides extends TileEntity {
  protected def SideCount = RotationHelper.getNumDirections

  protected def defaultState: Boolean = false

  var openSides = Array.fill(SideCount)(defaultState)

  def compressSides = (Direction.values(), openSides).zipped.foldLeft(0)((acc, entry) => acc | (if (entry._2) 1 << entry._1.ordinal() else 0)).toByte

  def uncompressSides(byte: Byte) = Direction.values().map(d => ((1 << d.ordinal()) & byte) != 0)

  def isSideOpen(side: Direction) = side != null && openSides(side.ordinal())

  def setSideOpen(side: Direction, value: Boolean): Unit = if (side != null && openSides(side.ordinal()) != value) {
    openSides(side.ordinal()) = value
  }

  override def loadForServer(nbt: CompoundNBT) {
    super.loadForServer(nbt)
    if (nbt.contains(Settings.namespace + "openSides"))
      openSides = uncompressSides(nbt.getByte(Settings.namespace + "openSides"))
  }

  override def saveForServer(nbt: CompoundNBT) {
    super.saveForServer(nbt)
    nbt.putByte(Settings.namespace + "openSides", compressSides)
  }

  @OnlyIn(Dist.CLIENT)
  override def loadForClient(nbt: CompoundNBT) {
    super.loadForClient(nbt)
    openSides = uncompressSides(nbt.getByte(Settings.namespace + "openSides"))
  }

  override def saveForClient(nbt: CompoundNBT) {
    super.saveForClient(nbt)
    nbt.putByte(Settings.namespace + "openSides", compressSides)
  }
}

File: src/main/scala/li/cil/oc/common/tileentity/traits/OpenSides.java
package li.cil.oc.common.tileentity.traits;

import li.cil.oc.integration.util.Wrench;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.nbt.NBTTagByte;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraftforge.common.util.INBTSerializable;

import javax.annotation.Nullable;
import java.util.Arrays;

/**
 * @author Vexatos, Sangar
 */
public final class OpenSides implements INBTSerializable<NBTTagByte> {
    public interface OpenSidesHost {
        void onSideOpenChanged(final EnumFacing side);
    }

    // ----------------------------------------------------------------------- //
    // Persisted data.

    private final boolean[] openSides = new boolean[EnumFacing.VALUES.length];

    // ----------------------------------------------------------------------- //
    // Computed data.

    private final OpenSidesHost host;

    // ----------------------------------------------------------------------- //

    public OpenSides(final OpenSidesHost host, final boolean defaultState) {
        this.host = host;
        Arrays.fill(openSides, defaultState);
    }

    public OpenSides(final OpenSidesHost host) {
        this(host, false);
    }

    public boolean isSideOpen(@Nullable final EnumFacing side) {
        return side != null && openSides[side.ordinal()];
    }

    public void setSideOpen(final EnumFacing side, final boolean value) {
        if (value != openSides[side.ordinal()]) {
            openSides[side.ordinal()] = value;
            host.onSideOpenChanged(side);
        }
    }

    public void toggleSide(final EnumFacing side) {
        openSides[side.ordinal()] = !openSides[side.ordinal()];
        host.onSideOpenChanged(side);
    }

    public boolean onActivated(final EntityPlayer player, final EnumHand hand, final BlockPos pos, final EnumFacing side) {
        if (Wrench.holdsApplicableWrench(player, pos)) {
            if (player.isServerWorld()) {
                final EnumFacing sideToToggle = player.isSneaking() ? side.getOpposite() : side;
                toggleSide(sideToToggle);
            }
            return true;
        }

        return false;
    }

    // ----------------------------------------------------------------------- //
    // INBTSerializable

    @Override
    public NBTTagByte serializeNBT() {
        int compressed = 0;
        for (int i = 0; i < openSides.length; i++) {
            if (openSides[i]) {
                compressed |= 1 << i;
            }
        }
        return new NBTTagByte((byte) compressed);
    }

    @Override
    public void deserializeNBT(final NBTTagByte nbt) {
        final int compressed = nbt.getByte() & 0xFF;
        for (int i = 0; i < openSides.length; i++) {
            openSides[i] = (compressed & (1 << i)) != 0;
        }
    }
}
